File: java/structures/LinkedLists/DoublyLinkedList.java
package structures.LinkedLists;

public class DoublyLinkedList<T> implements Iterable<T> {

  private int size = 0;
  private Node<T> head = null;
  private Node<T> tail = null;

  private static class Node<T> {
    T data;
    Node<T> prev, next;

    public Node(T data, Node<T> prev, Node<T> next) {
      this.data = data;
      this.prev = prev;
      this.next = next;
    }

    @Override
    public String toString() {
      return data.toString();
    }
  }

  public void clear() {
    Node<T> trav = head;
    while (trav != null) {
      Node<T> next = trav.next;
      trav.prev = trav.next = null;
      trav.data = null;
      trav = next;
    }
    head = tail = null;
    size = 0;
  }

  public int size() {
    return size;
  }

  public boolean isEmpty() {
    return size == 0;
  }

  public void add(T value) {
    addLast(value);
  }

  public void addFirst(T value) {
    if (isEmpty()) {
      head = tail = new Node<>(value, null, null);
    } else {
      Node<T> newNode = new Node<>(value, null, head);
      head.prev = newNode;
      head = newNode;
    }
    size++;
  }

  public void addLast(T value) {
    if (isEmpty()) {
      head = tail = new Node<>(value, null, null);
    } else {
      Node<T> newNode = new Node<>(value, tail, null);
      tail.next = newNode;
      tail = newNode;
    }
    size++;
  }

  public void addAt(int index, T value) {
    if (index < 0 || index > size)
      throw new IndexOutOfBoundsException("Index " + index + " is out of bounds.");

    if (index == 0) {
      addFirst(value);
      return;
    }

    if (index == size) {
      addLast(value);
      return;
    }

    Node<T> trav;
    if (index < size / 2) {
      trav = head;
      for (int i = 0; i < index - 1; i++)
        trav = trav.next;
    } else {
      trav = tail;
      for (int i = size - 1; i >= index; i--)
        trav = trav.prev;
    }

    Node<T> newNode = new Node<>(value, trav, trav.next);
    trav.next.prev = newNode;
    trav.next = newNode;
    size++;
  }

  public T getFirst() {
    if (isEmpty())
      throw new IllegalStateException("List is empty");
    return head.data;
  }

  public T getLast() {
    if (isEmpty())
      throw new IllegalStateException("List is empty");
    return tail.data;
  }

  public T removeFirst() {
    if (isEmpty())
      throw new IllegalStateException("List is empty");
    T data = head.data;
    head = head.next;
    size--;
    if (isEmpty())
      tail = null;
    else
      head.prev = null;
    return data;
  }

  public T removeLast() {
    if (isEmpty())
      throw new IllegalStateException("List is empty");
    T data = tail.data;
    tail = tail.prev;
    size--;
    if (isEmpty())
      head = null;
    else
      tail.next = null;
    return data;
  }

  private T remove(Node<T> node) {
    if (node.prev == null)
      return removeFirst();
    if (node.next == null)
      return removeLast();

    node.next.prev = node.prev;
    node.prev.next = node.next;
    T data = node.data;
    node.data = null;
    node.prev = node.next = null;
    size--;
    return data;
  }

  public T removeAt(int index) {
    if (index < 0 || index >= size)
      throw new IndexOutOfBoundsException("Index " + index + " is out of bounds.");

    Node<T> trav;
    if (index < size / 2) {
      trav = head;
      for (int i = 0; i < index; i++)
        trav = trav.next;
    } else {
      trav = tail;
      for (int i = size - 1; i > index; i--)
        trav = trav.prev;
    }
    return remove(trav);
  }

  public boolean remove(Object obj) {
    Node<T> trav = head;
    if (obj == null) {
      while (trav != null) {
        if (trav.data == null) {
          remove(trav);
          return true;
        }
        trav = trav.next;
      }
    } else {
      while (trav != null) {
        if (obj.equals(trav.data)) {
          remove(trav);
          return true;
        }
        trav = trav.next;
      }
    }
    return false;
  }

  public int indexOf(Object obj) {
    Node<T> trav = head;
    int index = 0;
    if (obj == null) {
      while (trav != null) {
        if (trav.data == null)
          return index;
        trav = trav.next;
        index++;
      }
    } else {
      while (trav != null) {
        if (obj.equals(trav.data))
          return index;
        trav = trav.next;
        index++;
      }
    }
    return -1;
  }

  public boolean contains(Object obj) {
    return indexOf(obj) != -1;
  }

  @Override
  public String toString() {
    StringBuilder sb = new StringBuilder("[");
    Node<T> trav = head;
    while (trav != null) {
      sb.append(trav.data);
      if (trav.next != null)
        sb.append(", ");
      trav = trav.next;
    }
    sb.append("]");
    return sb.toString();
  }

  @Override
  public java.util.Iterator<T> iterator() {
    return new java.util.Iterator<T>() {
      private Node<T> trav = head;

      @Override
      public boolean hasNext() {
        return trav != null;
      }

      @Override
      public T next() {
        T data = trav.data;
        trav = trav.next;
        return data;
      }
    };
  }

  // ---------------- MAIN ----------------
  public static void main(String[] args) {
    DoublyLinkedList<Integer> list = new DoublyLinkedList<>();
    System.out.println("Initial list: " + list);

    // Add elements
    list.addFirst(10);
    list.addLast(20);
    list.add(30);
    list.addAt(2, 49);
    System.out.println("After adds: " + list);

    // Access elements
    System.out.println("First: " + list.getFirst());
    System.out.println("Last: " + list.getLast());

    // Remove by index
    System.out.println("removeAt(1): " + list.removeAt(1));
    System.out.println("After removeAt: " + list);

    // Remove first/last
    System.out.println("removeFirst(): " + list.removeFirst());
    System.out.println("removeLast(): " + list.removeLast());
    System.out.println("After removeFirst/Last: " + list);

    // Add more and test remove(Object)
    list.add(100);
    list.add(200);
    list.add(300);
    System.out.println("Before remove(obj): " + list);
    list.remove((Integer) 200);
    System.out.println("After remove(200): " + list);

    // Iteration
    System.out.print("Iterating: ");
    for (int val : list)
      System.out.print(val + " ");
    System.out.println();

    // Clear
    list.clear();
    System.out.println("After clear, isEmpty: " + list.isEmpty());
  }
}

File: python/LinkedLists/DoublyLinkedList.py
from typing import Generic, TypeVar, Optional, Iterator

T = TypeVar('T')


class DoublyLinkedList(Generic[T]):
    class _Node:
        __slots__ = ('data', 'prev', 'next')

        def __init__(self, data: T, prev: Optional['DoublyLinkedList._Node'], next: Optional['DoublyLinkedList._Node']):
            self.data = data
            self.prev = prev
            self.next = next

        def __str__(self) -> str:
            return str(self.data)

    def __init__(self) -> None:
        self._size: int = 0
        self._head: Optional[DoublyLinkedList._Node] = None
        self._tail: Optional[DoublyLinkedList._Node] = None

    def clear(self) -> None:
        trav = self._head
        while trav is not None:
            next = trav.next
            trav.prev = trav.next = None
            trav.data = None  # type: ignore
            trav = next
        self._head = self._tail = None
        self._size = 0

    def size(self) -> int:
        return self._size

    def is_empty(self) -> bool:
        return self._size == 0

    def add(self, value: T) -> None:
        self.add_last(value)

    def add_first(self, value: T) -> None:
        if self.is_empty():
            node = DoublyLinkedList._Node(value, None, None)
            self._head = self._tail = node
        else:
            node = DoublyLinkedList._Node(value, None, self._head)
            self._head.prev = node  # type: ignore
            self._head = node
        self._size += 1

    def add_last(self, value: T) -> None:
        if self.is_empty():
            node = DoublyLinkedList._Node(value, None, None)
            self._head = self._tail = node
        else:
            node = DoublyLinkedList._Node(value, self._tail, None)
            self._tail.next = node  # type: ignore
            self._tail = node
        self._size += 1

    def add_at(self, index: int, value: T) -> None:
        if index < 0 or index > self._size:
            raise IndexError(f"Index {index} is out of bounds.")
        if index == 0:
            self.add_first(value)
            return
        if index == self._size:
            self.add_last(value)
            return
        if index < self._size // 2:
            trav = self._head
            for _ in range(index - 1):
                trav = trav.next  # type: ignore
        else:
            trav = self._tail
            for _ in range(self._size - 1, index - 1, -1):
                trav = trav.prev  # type: ignore
        new_node = DoublyLinkedList._Node(
            value, trav, trav.next)  # type: ignore
        trav.next.prev = new_node  # type: ignore
        trav.next = new_node  # type: ignore
        self._size += 1

    def get_first(self) -> T:
        if self.is_empty():
            raise IndexError("List is empty")
        return self._head.data  # type: ignore

    def get_last(self) -> T:
        if self.is_empty():
            raise IndexError("List is empty")
        return self._tail.data  # type: ignore

    def remove_first(self) -> T:
        if self.is_empty():
            raise IndexError("List is empty")
        data = self._head.data  # type: ignore
        self._head = self._head.next  # type: ignore
        self._size -= 1
        if self.is_empty():
            self._tail = None
        else:
            self._head.prev = None  # type: ignore
        return data

    def remove_last(self) -> T:
        if self.is_empty():
            raise IndexError("List is empty")
        data = self._tail.data  # type: ignore
        self._tail = self._tail.prev  # type: ignore
        self._size -= 1
        if self.is_empty():
            self._head = None
        else:
            self._tail.next = None  # type: ignore
        return data

    def _remove_node(self, node: _Node) -> T:
        if node.prev is None:
            return self.remove_first()
        if node.next is None:
            return self.remove_last()
        node.next.prev = node.prev
        node.prev.next = node.next
        data = node.data
        node.data = None  # type: ignore
        node.prev = node.next = None
        self._size -= 1
        return data  # type: ignore

    def remove_at(self, index: int) -> T:
        if index < 0 or index >= self._size:
            raise IndexError(f"Index {index} is out of bounds.")
        if index < self._size // 2:
            trav = self._head
            for _ in range(index):
                trav = trav.next  # type: ignore
        else:
            trav = self._tail
            for _ in range(self._size - 1, index, -1):
                trav = trav.prev  # type: ignore
        return self._remove_node(trav)  # type: ignore

    def remove(self, obj: object) -> bool:
        trav = self._head
        if obj is None:
            while trav is not None:
                if trav.data is None:
                    self._remove_node(trav)
                    return True
                trav = trav.next
        else:
            while trav is not None:
                if obj == trav.data:
                    self._remove_node(trav)
                    return True
                trav = trav.next
        return False

    def index_of(self, obj: object) -> int:
        trav = self._head
        idx = 0
        if obj is None:
            while trav is not None:
                if trav.data is None:
                    return idx
                trav = trav.next
                idx += 1
        else:
            while trav is not None:
                if obj == trav.data:
                    return idx
                trav = trav.next
                idx += 1
        return -1

    def contains(self, obj: object) -> bool:
        return self.index_of(obj) != -1

    def __str__(self) -> str:
        sb = []
        trav = self._head
        while trav is not None:
            sb.append(str(trav.data))
            trav = trav.next
        return "[" + ", ".join(sb) + "]"

    def __iter__(self) -> Iterator[T]:
        trav = self._head
        while trav is not None:
            yield trav.data  # type: ignore
            trav = trav.next


# ---------------- MAIN ----------------
if __name__ == "__main__":
    lst = DoublyLinkedList[int]()
    print("Initial list:", lst)

    # Add elements
    lst.add_first(10)
    lst.add_last(20)
    lst.add(30)
    lst.add_at(2, 49)
    print("After adds:", lst)

    # Access elements
    print("First:", lst.get_first())
    print("Last:", lst.get_last())

    # Remove by index
    print("remove_at(1):", lst.remove_at(1))
    print("After remove_at:", lst)

    # Remove first/last
    print("remove_first():", lst.remove_first())
    print("remove_last():", lst.remove_last())
    print("After remove_first/last:", lst)

    # Add more and test remove(obj)
    lst.add(100)
    lst.add(200)
    lst.add(300)
    print("Before remove(obj):", lst)
    lst.remove(200)
    print("After remove(200):", lst)

    # Iteration
    print("Iterating:", end=" ")
    for val in lst:
        print(val, end=" ")
    print()

    # Clear
    lst.clear()
    print("After clear, is_empty:", lst.is_empty())
